architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,239 @@
+2. Secrets managed by Radius are always placed in the same namespace as the object that "owns" them. This is a requirement of the Kubernetes RBAC model.
+3. Secrets managed by Radius are subject to the Kubernetes RBAC model for controlling access. Operators are expected to limit access for users using existing tools.
+
+**Status:** All mitigations listed are currently active. Operators are expected to secure their cluster and limit access for users.
+
+#### Escalation of Privilege by Using Radius to Circumvent Kubernetes RBAC Controls
+
+**Description:** A malicious actor could circumvent Kubernetes RBAC controls and create arbitrary resources in Kubernetes by using the `Recipe` custom resource.
+
+The `Recipe` controller has limited permissions, so it cannot be used directly to escalate privileges in Kubernetes. However, it calls into UCP/Radius, which operates with a wide scope of permissions in Kubernetes and the cloud.
+
+Authorized users with access to create a `Recipe` resource in Kubernetes can execute any Recipe in any Environment registered with Radius.
+
+At the time of writing, Radius does not provide granular authorization controls. Any authenticated client can create any Radius resource and execute any action Radius is capable of taking. This is not limited to the Kubernetes controllers.
+
+**Impact:** An authorized user of the Kubernetes cluster with permission to create a `Recipe` resource can execute any Recipe in any Environment registered with Radius.
+
+**Mitigations:**
+
+1. Operators should limit access to the `Recipe` resource using Kubernetes RBAC.
+2. Operators should limit direct access to the Radius API using Kubernetes RBAC.
+3. We should revisit the threat model and provide a more robust set of authorization controls when granular authorization policies are added to Radius.
+
+**Status:** These mitigations are partial and require configuration by the operator. We will revisit and improve this area in the future.
+
+## Open Questions
+
+## Action Items
+
+1. Use a hashing algorithm other than SHA-1 while computing the hash of the configuration of a Deployment object. This is a breaking change because deployments that are already hashed with SHA1 should be redeployed so that reconciler can work as expected.
+2. Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function. Make changes to the necessary components if required.
+3. Define and implement necessary Network Policies to ensure that communication is accepted only from expected and authorized components. Regularly review and update these policies to maintain security.
+4. Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root. Do the necessary updates.
```

Comment:
I'd prefer that these "review" items be handled as part of the threat model review.

For example:

> Check if RBAC with Least Privilege is configured for every component to ensure that each component has only the permissions it needs to function.

The threat model is the **right place for us to define** the "permissions each component needs to function". It doesn't make sense to push this work out of the threat model. 

---

> Containers should run as a non-root user wherever possible to minimize the risks. Check if we can run any of the Radius containers as non-root.

This is just scratching the surface in terms of policies applied to our components 😆. See: https://kubernetes.io/docs/concepts/security/pod-security-standards/#policy-instantiation